File: blas/cpp/int64_arm64.cpp
#include <arm_neon.h>
#include <stdint.h>

void add_int64(int64x2_t *input1, int64x2_t *input2, int64x2_t *output, uint64_t size) {
    int chunks = size / 2;

    for (int i = 0; i < chunks; i++) {
        output[i] = vaddq_s64(input1[i], input2[i]);
    }
    // Leftovers
    for (int i = chunks * 2; i < size; i++) {
        ((int64_t *) output)[i] = ((int64_t *) input1)[i] + ((int64_t *) input2)[i];
    }
}

void sub_int64(int64x2_t *input1, int64x2_t *input2, int64x2_t *output, uint64_t size) {
    int chunks = size / 2;

    for (int i = 0; i < chunks; i++) {
        output[i] = vsubq_s64(input1[i], input2[i]);
    }
    // Leftovers
    for (int i = chunks * 2; i < size; i++) {
        ((int64_t *) output)[i] = ((int64_t *) input1)[i] - ((int64_t *) input2)[i];
    }
}

void mul_int64(int64x2_t *input1, int64x2_t *input2, int64x2_t *output, uint64_t size) {
    int chunks = size / 2;

//    for (int i = 0; i < chunks; i++) {
//        output[i] = vmulq_s64(input1[i], input2[i]);
//    }
    // Leftovers
    chunks = 0;
    for (int i = chunks * 2; i < size; i++) {
        ((int64_t *) output)[i] = ((int64_t *) input1)[i] * ((int64_t *) input2)[i];
    }
}

File: blas/int64_arm64.go
package blas

import "unsafe"

//go:noescape

func _add_int64(input1, input2, output unsafe.Pointer, size uint64)
func (o Int64s) Add(input1, input2 []int64) {
	_add_int64(unsafe.Pointer(&input1[0]), unsafe.Pointer(&input2[0]), unsafe.Pointer(&(o)[0]), uint64(len(o)))
}

func _sub_int64(input1, input2, output unsafe.Pointer, size uint64)
func (o Int64s) Sub(input1, input2 []int64) {
	_sub_int64(unsafe.Pointer(&input1[0]), unsafe.Pointer(&input2[0]), unsafe.Pointer(&(o)[0]), uint64(len(o)))
}

func _mul_int64(input1, input2, output unsafe.Pointer, size uint64)
func (o Int64s) Mul(input1, input2 []int64) {
	_mul_int64(unsafe.Pointer(&input1[0]), unsafe.Pointer(&input2[0]), unsafe.Pointer(&(o)[0]), uint64(len(o)))
}
